content/exploratory_notebooks/facebook_notebook.md

```diff
@@ -0,0 +1,463 @@
+ 
+ The number of unique triangles in the network are found next:
+
+```{code-cell} ipython3
+sum(list(nx.triangles(G).values())) / 3  # divide by 3 because each triangle is counted once for each node
+```
+
+ Now the average number of triangles that a node is a part of:
+
+```{code-cell} ipython3
+np.mean(list(nx.triangles(G).values()))
+```
+
+Due to having some nodes that belong to a great many triangles, the metric of median will give us a better understanding:
+
+```{code-cell} ipython3
+np.median(list(nx.triangles(G).values()))
+```
+
+In fact, the median value is just $161$ triangles, when the mean is around $1197$ triangles that a node is part of. That means that the majority of nodes of the network belong to extremely few triangles, whereas some nodes are part of a plethora of triangles (which are extreme values that increase the mean)
+
+In conclusion, the high average clustering coefficient together with the huge number of triangles are signs of the triadic closure. In detail, the triadic closure means that as time goes on, new edges tend to form between two users that have one or more mutual friends. That can be explained by the fact that Facebook usually suggests new friends to a user when there are many mutual friends between the user and the new friend to be added. Also, there is a source of latent stress. For example, if node $A$ is friends with
+node $B$ and $C$,  some tension builds up if $B$ and $C$ are not friends with each other.
+
++++
+
+## Bridges
+First of all, an edge joining two nodes A and B in the graph is considered a bridge, if deleting the edge would cause A and B to lie in two different components. Now it is checked if there are any bridges in this network:
+
+```{code-cell} ipython3
+nx.has_bridges(G)
+```
+
+Actually, there are bridges in the network. Now the edges that are bridges will be saved in a list and the number of them is printed:
+
+```{code-cell} ipython3
+bridges = []
+for edge in nx.bridges(G):
+    bridges.append(edge)
+len(bridges)
+```
+
+The existence of so many bridges is due to the fact that this network only contains the spotlight nodes and the friends of them. As a result, some friends of spotlight nodes are only connected to a spotlight node, making that edge a bridge.
+
+Also, the edges that are local bridges are saved in a list and their number is printed. In detaill, an edge joining two nodes $C$ and $D$ 
+in a graph is a local bridge, if its endpoints $C$ and $D$ have no friends in common. Very importantly, an edge that is a bridge is also a local bridge. Thus, this list contains all the above bridges as well:
+
+```{code-cell} ipython3
+local_bridges = []
+for edge in nx.local_bridges(G):
+    local_bridges.append(eval( '(' + str(edge[0]) + '),(' + str(edge[1]) + ')' ))  # save the local bridge as a tuple
```

Comment:
```suggestion
local_bridges = list(nx.local_bridges(G, with_span=False))
```
Similar suggestion as above. The `with_span` is a more concise way limiting the output to the edge itself (similar to what the original code was aiming to do with the `eval`.